### PR TITLE
Quantile calculation with index

### PR DESCRIPTION
**Pull request by warrickpk**
Changes to the quantile calculators for calculating VaR and ES.
The changes allow QuantileResult to be used to describe the results from a QuantileCalculationMethod. This allows for value, calculation index and weight statistics to be output from any quantile calculation. The calculation index is an integer index array of data points used to calculate the quantile statistic.

### CODE REVIEW

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/QuantileCalculationMethod.java [hunk 1]
```diff
@@ -63,34 +43,14 @@ public double quantileFromUnsorted(double level, DoubleArray sample) {
...
-   * <p>
-   * The quantile level is in decimal, i.e. 99% = 0.99 and 0 < level < 1 should be satisfied.
-   * This is measured from the bottom, that is, the quantile estimation with the level 99% corresponds to 
-   * the smallest 99% observations and 1% of the observation are above that level.
-   * <p>
-   * If index value computed from the level is outside of the sample data range, the nearest data point is used, i.e., 
-   * quantile is computed with flat extrapolation. 
-   * <p> 
-   * The sample observations are supposed to be unsorted, the first step is to sort the data.
+   * The sample observations are supposed to be unsorted.
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/QuantileCalculationMethod.java]: Please add javadoc mentioning that `indices` in `QuantileResult` are based on the original unsorted array.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/QuantileCalculationMethod.java [hunk 2]
```diff
@@ -103,37 +63,16 @@ public double quantileWithExtrapolationFromUnsorted(double level, DoubleArray sa
...
-   * The quantile level is in decimal, i.e. 99% = 0.99 and 0 < level < 1 should be satisfied.
-   * This is measured from the bottom, that is, Thus the expected shortfall with the level 99% corresponds to 
-   * the average of the smallest 99% of the observations.
-   * <p>
-   * If index value computed from the level is outside of the sample data range, the nearest data point is used, i.e., 
-   * expected short fall is computed with flat extrapolation.
    * Thus this is coherent to {@link #quantileWithExtrapolationFromUnsorted(double, DoubleArray)}.
    * <p> 
-   * The sample observations are supposed to be unsorted, the first step is to sort the data.
+   * The sample observations are sorted from the smallest to the largest.
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/QuantileCalculationMethod.java]: Is this the correct comment for the unsorted method?

Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/QuantileCalculationMethod.java]: Please add javadoc mentioning that `indices` in `QuantileResult` are based on the original unsorted array.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/QuantileCalculationMethod.java [hunk 3]
```diff
@@ -23,34 +23,14 @@
...
-   * <p>
-   * The quantile level is in decimal, i.e. 99% = 0.99 and 0 < level < 1 should be satisfied.
-   * This is measured from the bottom, that is, the quantile estimation with the level 99% corresponds to 
-   * the smallest 99% observations and 1% of the observation are above that level.
-   * <p>
-   * If index value computed from the level is outside of the sample data range, 
-   * {@code IllegalArgumentException} is thrown. 
-   * <p> 
-   * The sample observations are supposed to be unsorted, the first step is to sort the data.
+   * The sample observations are supposed to be unsorted.
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/QuantileCalculationMethod.java]: Please add javadoc mentioning that `indices` in `QuantileResult` are based on the original unsorted array.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/DiscreteQuantileMethod.java [hunk 4]
```diff
@@ -6,66 +6,99 @@
...
    * @return the index in the sample
    */
   abstract int index(double quantileSize);
 
   /**
    * Internal method returning the sample size correction for the specific implementation.
-   * 
-   * @param sampleSize  the sample size
+   *
+   * @param sampleSize the sample size
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/DiscreteQuantileMethod.java]: Two spaces between parameter name and comment.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/DiscreteQuantileMethod.java [hunk 5]
```diff
@@ -6,66 +6,99 @@
...
    * @return the index shift
    */
   abstract double indexShift();
+
+  /**
+   * Generate an index of doubles.
+   * <p>
+   * Creates an index of doubles from 1.0 to a stipulated number, in increments of 1.
+   *
+   * @param indexArrayLength length of index array to be created
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/DiscreteQuantileMethod.java]: Two spaces between parameter name and comment.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/DiscreteQuantileMethod.java [hunk 6]
```diff
@@ -6,66 +6,99 @@
...
+   * Shift added to/subtracted from index during intermediate steps in the expected shortfall computation.
+   *
    * @return the index shift
    */
   abstract double indexShift();
+
+  /**
+   * Generate an index of doubles.
+   * <p>
+   * Creates an index of doubles from 1.0 to a stipulated number, in increments of 1.
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/DiscreteQuantileMethod.java]: Is this from 0.0?

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java [hunk 7]
```diff
@@ -8,64 +8,60 @@
...
 
   /**
-   * Constructor. 
+   * Constructor.
    * <p>
    * The exponential weight lambda must be > 0 and < 1.0.
-   * 
-   * @param lambda  the exponential weight
+   *
+   * @param lambda the exponential weight
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java]: Two spaces between parameter name and comment.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java [hunk 8]
```diff
@@ -8,64 +8,60 @@
...
+   * The details consists on the indices of the samples actually used in the quantile computation - indices in the
+   * input sample - and the weights for each of those samples. The details are sufficient to recompute the
    * quantile directly from the input sample.
    * <p>
    * The sample observations are supposed to be unsorted, the first step is to sort the data.
-   * 
-   * @param level  the quantile level
-   * @param sample  the sample observations
+   *
+   * @param level the quantile level
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java]: Two spaces between parameter name and comment.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java [hunk 9]
```diff
@@ -8,64 +8,60 @@
...
+   * @param sample the sample observations
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java]: Two spaces between parameter name and comment.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java [hunk 10]
```diff
@@ -76,31 +72,31 @@ public QuantileResult quantileDetailsFromUnsorted(double level, DoubleArray samp
...
+   * in the input sample - and the weights for each of those samples. The details are sufficient to recompute the
    * expected shortfall directly from the input sample.
-   * <p> 
+   * <p>
    * The sample observations are supposed to be unsorted, the first step is to sort the data.
-   * 
-   * @param level  the quantile level
-   * @param sample  the sample observations
+   *
+   * @param level the quantile level
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java]: Two spaces between parameter name and comment.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java [hunk 11]
```diff
@@ -76,31 +72,31 @@ public QuantileResult quantileDetailsFromUnsorted(double level, DoubleArray samp
...
+   * @param sample the sample observations
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java]: Two spaces between parameter name and comment.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java [hunk 12]
```diff
@@ -130,18 +126,19 @@ private QuantileResult quantileDetails(
...
    * Computes value-at-risk.
-   * @param index  the index from which the VaR should be computed
-   * @param runningWeight  the running weight up to index
-   * @param isExtrapolated  flag indicating if value should be extrapolated (flat) beyond the last value
-   * @param s  the sorted sample
-   * @param w  the sorted weights
-   * @param order  the order of the sorted sample in the unsorted sample
-   * @param level  the level at which the VaR should be computed
+   *
+   * @param index the index from which the VaR should be computed
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java]: Two spaces between parameter name and comment.
The same comments for the other parameters.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java [hunk 13]
```diff
@@ -160,30 +157,30 @@ private QuantileResult quantileFromIndexRunningWeight(
...
    * Computes expected shortfall.
-   * @param index  the index from which the ES should be computed
-   * @param runningWeight  the running weight up to index
-   * @param isExtrapolated  flag indicating if value should be extrapolated (flat) beyond the last value
-   * @param s  the sorted sample
-   * @param w  the sorted weights
-   * @param order  the order of the sorted sample in the unsorted sample
-   * @param level  the level at which the ES should be computed
+   *
+   * @param index the index from which the ES should be computed
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java]: Two spaces between parameter name and comment.
The same comments for the other parameters.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java [hunk 14]
```diff
@@ -202,22 +199,23 @@ private QuantileResult esFromIndexRunningWeight(
...
+  protected QuantileResult expectedShortfall(double level, DoubleArray sortedSample) {
     throw new UnsupportedOperationException("Expected Shortfall only from unsorted sample due to weights.");
   }
-  
+
   /**
    * Returns the weights for a given sample size.
-   * @param size  the sample size
+   *
+   * @param size the sample size
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/ExponentiallyWeightedInterpolationQuantileMethod.java]: Two spaces between parameter name and comment.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/InterpolationQuantileMethod.java [hunk 15]
```diff
@@ -6,75 +6,114 @@
...
    * @return the correction
    */
   abstract double indexCorrection();
 
   /**
    * Internal method returning the sample size correction for the specific implementation.
-   * 
-   * @param sampleSize  the sample size
+   *
+   * @param sampleSize the sample size
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/InterpolationQuantileMethod.java]: Two spaces between parameter name and comment.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/InterpolationQuantileMethod.java [hunk 16]
```diff
@@ -6,75 +6,114 @@
...
+    double[] s = sample.toArray();
+    DoubleArrayMath.sortPairs(s, order);
     int lowerIndex = (int) Math.floor(adjustedLevel);
     int upperIndex = (int) Math.ceil(adjustedLevel);
     double lowerWeight = upperIndex - adjustedLevel;
     double upperWeight = 1d - lowerWeight;
-    return lowerWeight * sortedSample.get(lowerIndex - 1) + upperWeight * sortedSample.get(upperIndex - 1);
+    return QuantileResult.of(
+        lowerWeight * s[lowerIndex - 1] + upperWeight * s[upperIndex - 1],
+        new int[]{(int) order[(int) Math.floor(adjustedLevel) - 1], (int) order[(int) Math.ceil(adjustedLevel) - 1]},
```
Comment (yukiiwashita) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/InterpolationQuantileMethod.java]: `(int) Math.floor(adjustedLevel)` and `(int) Math.ceil(adjustedLevel)` are already computed in 35 and 36, respectively.

File: modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/QuantileCalculationMethod.java [hunk 17]
```diff
@@ -23,34 +23,16 @@
    * If index value computed from the level is outside of the sample data range, 
    * {@code IllegalArgumentException} is thrown. 
    * <p> 
-   * The sample observations are sorted from the smallest to the largest. 
-   * 
-   * @param level  the quantile level
-   * @param sortedSample  the sample observations
-   * @return the quantile estimation
-   */
-  public double quantileFromSorted(double level, DoubleArray sortedSample) {
```
Comment (marc-henrard) [inline modules/math/src/main/java/com/opengamma/strata/math/impl/statistics/descriptive/QuantileCalculationMethod.java]: We probably want to keep the existing methods and add new ones. Beyond backward compatibility, there is value in keeping those simpler methods returning simply a double (from sorted or notSorted). I would not remove the exiting methods.

**Pull request merged**